Routing: Standard-behavior for components

Hackwar opened the pull request:
This PR implements a generic rule for standard behavior of components which routers extend from JComponentRouterAdvanced.
### How to test
- Apply this change
- Change the router of com_content by replacing the code with the following:
  
  ``` php
  <?php
  class ContentRouter extends JComponentRouterAdvanced
  {
      public function __construct($app = null, $menu = null) {
  
          $this->registerView('categories', 'categories');
          $this->registerView('category', 'category', 'id', 'categories', '', true, array('default', 'blog'));
          $this->registerView('article', 'article', 'id', 'category', 'catid');
          $this->registerView('archive', 'archive');
          $this->registerView('featured', 'featured');
          $this->registerView('form', 'form');
          parent::__construct($app, $menu);
  
          $this->attachRule(new JComponentRouterRulesMenu($this));
          $this->attachRule(new JComponentRouterRulesNomenu($this));
          $this->attachRule(new JComponentRouterRulesStandard($this));
      }
  
      public function getArticleKey($segments, $vars)
      {
          return (int) $segments;
      }
  }
  ```
- Comment line 69 and 123 from /components/com_content/helpers/route.php to disable the lookup of the Itemid there.
- See that it creates sane URLs and that those URLs point to the right place.
### Issues

This requires that #5446 and all depending requests are merged first. (You can still test this PR, since all changes from those PRs are included into this one.)

This was made possible through the generous donation of the people mentioned in the following link via an Indiegogo campaign: http://joomlager.de/crowdfunding/5-contributors

Hackwar committed:
Implementing JComponentRouterAdvanced class

Hackwar committed:
Implementing JComponentRouterAdvanced class. Fixes #5444

Call-time pass-by-reference has been removed

Removing JError, using Exception instead

protecting $name and renaming register() to registerView()

Adding removeRule, getRules and renamed $id to $key in register method

Making method names consistent

Implementing JComponentRouterViewconfiguration for configuration of views in JComponentRouterAdvanced

Codestyle, smaller improvements, unittests for all component router classes except for JComponentRouterAdvanced

Removing ability to have one view with different names and implementing unittests for JComponentRouterAdvanced

Adding get<View>Slug() and get<View>Key() methods to JComponentRouterAdvanced

Updating unittest

Small fixes

Adding back in platform check

Adding back in platform check

Adding back in platform check

Adding back in platform check

Implementing feedback so far

Adding "covers" notation for unittests

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Implementing JComponentRouterRulesMenu class to find right Itemid

Hackwar committed:
Call-time pass-by-reference has been removed

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Removing JError, using Exception instead

Hackwar committed:
protecting $name and renaming register() to registerView()

Hackwar committed:
Codestyle and minor bug

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Adding removeRule, getRules and renamed $id to $key in register method

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Implementing legacy router for com_content

Hackwar committed:
Codestylefix

zero-24 committed:
Fix CS issues that was found by Travis

https://travis-ci.org/joomla/joomla-cms/jobs/44966357

Hackwar committed:
Merge pull request #14 from zero-24/patch-25

Fix CS issues that was found by Travis

Hackwar committed:
Changing component router to retrieve, since the com_content router is pretty complex and throws errors in our testing environment

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Hackwar committed:
Merge branch 'com_content_router_legacy' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Hackwar committed:
Making method names consistent

mbabker committed:
Change targetplatform version check to match full CMS version Fixes #5546

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Implementing JComponentRouterViewconfiguration for configuration of views in JComponentRouterAdvanced

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Codestyle, smaller improvements, unittests for all component router classes except for JComponentRouterAdvanced

Hackwar committed:
Removing ability to have one view with different names and implementing unittests for JComponentRouterAdvanced

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Fixes to JComponentRouterRulesMenu

Hackwar committed:
Small fixes and unittests for JComponentRouterRulesMenu

Hackwar committed:
Codestyle

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Hackwar committed:
Implementing new view registering method in com_content

Hackwar committed:
Codestyle

Hackwar committed:
Implementing "no Itemid" behavior

Hackwar committed:
Unittest improvements

Hackwar committed:
Implementing standard routing behavior for components

Hackwar committed:
Adding get<View>Slug() and get<View>Key() methods to JComponentRouterAdvanced

Hackwar committed:
Updating unittest

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesstandard

Hackwar committed:
Small fixes

Hackwar committed:
Adding back in platform check

Hackwar committed:
Adding back in platform check

Hackwar committed:
Adding back in platform check

Hackwar committed:
Adding back in platform check

Hackwar committed:
Codestyle

Hackwar committed:
Codestyle

joomdonation commented on libraries/cms/component/router/rules/standard.php:
```diff
@@ -0,0 +1,272 @@
+		$views = $this->router->getViews();
+
+		// Return directly when the URL of the Itemid is identical with the URL to build
+		if (isset($item->query['view']) && isset($query['view']) && $item->query['view'] == $query['view'])
```
I think we can remove isset($query['view']) from this line because you have a check for it in line 179 already.

Hackwar replied:
You are right. Fixed it

joomdonation replied:
Thanks Hannes!

Hackwar committed:
Removing unnecessary check

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrouteradvanced

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrulesitemid

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into com_content_router_legacy

Hackwar committed:
Implementing feedback so far

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrouteradvanced

Hackwar committed:
Adding "covers" notation for unittests

Hackwar committed:
Merge branch 'componentrouteradvanced' of github.com:Hackwar/joomla-cms into componentrulesitemid

Hackwar committed:
Updating to latest changes to JComponentRouterView

Hackwar committed:
Merge branch '3.5-dev' of https://github.com/joomla/joomla-cms into componentrulesitemid

Conflicts:
	tests/unit/suites/libraries/cms/component/router/JComponentRouterViewTest.php

JoshuaLewis commented:
Any news about this?

Commit:
Joomla! Update notification email plugin. Fixes #6886.

wilsonge committed:
Merge branch '6886' into 3.5-dev

wilsonge committed:
Merge branch 'staging' into 3.5-dev

Commit:
Merge branch 'staging' into 3.5-dev

pe7er committed:
[imp] Adds Item count to Category Manager. Closes #6916.

wilsonge committed:
Merge branch 'staging' into 3.5-dev

wilsonge committed:
Update to String package 1.3. Fixes #6600.

PHP7 fixes

zero-24 committed:
update mail plugin CS

zero-24 committed:
router cs

zero-24 committed:
Fix JURI to JUri

zero-24 committed:
fix spaces

Commit:
Missing file from 834ad2c6d3cf1819a2479a451723e52cd26b1886 (Fix #6989)

Commit:
Fix unit test failure in JComponentRouterViewTest

zero-24 committed:
add the new plugin to the script.php

zero-24 committed:
typo

pe7er committed:
[imp] Category Manager added "Count Items" for Contacts. Closes #6938.

pe7er committed:
[imp] Category Manager added "Count Items" for Banners. Standardises to filter.published. Closes #6934.

pe7er committed:
[imp] Category Manager added "Count Items" for NewsFeeds. Closes #6936.

wilsonge committed:
Merge pull request #6990 from mbabker/35testFix

Fix unit test failure in JComponentRouterViewTest

Hackwar committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into componentrulesitemid

Hackwar committed:
Merge branch '3.5-dev' of https://github.com/joomla/joomla-cms into componentrulesitemid

Hackwar committed:
Fixing unittests of JComponentRouterRulesMenuTest

wilsonge committed:
Merge pull request #6991 from zero-24/addnewplugintoscript

[3.5-dev] Add the new plugin to the script.php

wilsonge committed:
Merge pull request #6988 from zero-24/cs-35-dev

[3.5-dev] Fix some codestyle for the current 3.5-dev changes.

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into componentrulesnomenu2

Hackwar committed:
Fixing unittests of JComponentRouterRulesNomenu and adapting to JComponentRouterView

Hackwar committed:
Merge branch 'componentrulesitemid' of github.com:Hackwar/joomla-cms into com_content_router_legacy

Conflicts:
	components/com_content/helpers/route.php

Hackwar committed:
Updating the legacy router of com_content to use JComponentRouterView

mbabker committed:
[imp] Rename the install app JHtml helper. Closes #7017.

Hackwar committed:
Improving the router

wilsonge committed:
Merge branch 'staging' into 3.5-dev

Commit:
Added login background setting.

Fixes #6703

Buddhima committed:
Added Flip function to JImage.

Fixes #6571

Mathewlenning committed:
Code cleanup of JArchiveGzip.

Fixes #6497

wilsonge committed:
Merge branch 'staging' into 3.5-dev

infograf768 committed:
Add purge toolbar button in lang overrides manager

infograf768 committed:
[imp] Multilanguage: user site language should also be a published content language

joomdonation committed:
Make it easier to search by id and username in com_users backend.

Fixes #6755.

Commit:
UI: Adding level indicators for JFormFieldMenuItem.

Fixes #6782

wilsonge committed:
Merge branch 'staging' into 3.5-dev

dgrammatiko committed:
Load com_installer scripts through the JDocument API

Fixes #6706

Hackwar committed:
Merge branch 'com_content_router_legacy' of github.com:Hackwar/joomla-cms into componentrulesstandard

Hackwar committed:
Updateing JComponentRouterRulesStandard to use JComponentRouterView

Hackwar committed:
Merge branch 'componentrulesstandard' of github.com:Hackwar/joomla-cms into componentrulesstandard

maxvalentini77 committed:
[imp] Requiring a minimum of 1 for session lifetime. Adds a new JFormRuleNumber. Closes #6954.

wilsonge committed:
Merge branch 'staging' into 3.5-dev

dgrammatiko committed:
Move ISIS inline script to template.js

Fixes #6696

Commit:
Fix method declarations of test stubs

Commit:
Fix one more declaration

Commit:
Declare method static

mbabker committed:
Declare mockGet static

mbabker committed:
Ensure all mock classes have the methods declared static

mbabker committed:
Fixes in JCacheConstructTest

wilsonge committed:
Make mock function static

mbabker committed:
Test fixes for PHP 7

mbabker committed:
Restructure JErrorPage to support PHP 7 exceptions

mbabker committed:
More test fixes

mbabker committed:
Remove tests that are testing typehints

wilsonge committed:
Merge pull request #7065 from mbabker/PHP7ErrorPage

Restructure JErrorPage to support PHP 7 exceptions

mbabker committed:
Change patch content to use a variable since JFile::write uses the buffer by reference

wilsonge committed:
Merge branch 'staging' into 3.5-dev

mbabker committed:
Port https://github.com/joomla/joomla-framework/commit/9845075747c287cb7a71bcdd5616e1a9d2f05f15 for JArrayHelperTest

mbabker committed:
Merge branch 'staging' into 3.5-dev

mbabker committed:
Test fix for JForm

zero-24 committed:
fix the plugin name

Hackwar committed:
Merge branch 'componentrulesnomenu' of github.com:Hackwar/joomla-cms into componentrulesstandard

dgrammatiko committed:
Messages modal converted to Bootstrap Fixes #4563

Commit:
Merge pull request #7083 from zero-24/pluginfix

[3.5-dev] Lets the name lowercase as all other core extensions

Commit:
Merge branch 'staging' into 3.5-dev

maxvalentini77 committed:
[fix] Set "since" tag to 3.5.0 for new JFormRuleNumber. Fixes #7109

zero-24 committed:
Add the new language files to the install xml

wilsonge committed:
Merge pull request #7125 from zero-24/language_fix_update_notification

Add the new language files to the install xml

wilsonge committed:
Merge branch 'staging' into 3.5-dev

geejay101 committed:
[fix] User Admin with editstate permission cannot block/unblock users. Fixes #7134.

Commit:
Merge branch 'staging' into 3.5-dev

sion cannot block/unblock users

http://issues.joomla.org/tracker/joomla-cms/7131

Problem:

It is not possible to create admin groups which have the ability to deactivate / activate other users, but not change any other user rights. An administrator/user with only EditState permission for the Administration User component cannot deactivate / activate other users. When clicking on block/unblock in the userlist, the status of the user doesnt change.

The problem occurs because the tick-boxes in the userlist are not shown when the administrator has only EditState permission. The tickbox is only shown when the administrator has also Edit permission.

Solution:
administrator/components/com_users/views/users/tmpl/default.php
must be patched in line 94:
old code in Joomla 3.4.1: <?php if ($canEdit) : ?>
New: <?php if ($canEdit || $canChange) : ?>

#### Steps to reproduce the issue

Create a user / group so that the user has the following calculated settings in the user com_user administration component.

Configure ACL & Options: Not Allowed
Access Administration Interface: Allowed
Create: Allowed
Delete: Not Allowed.
Edit: Not Allowed.
Edit: State Allowed

Log into the administrator backend with that user and access the user administration.

#### Expected result

The user should be able to block/unblock other users with the switches in the "enabled" column

#### Actual result

The user can see the block/unblock switches in the "Enabled" column but when clicking the status of the user is not changing.

The problem occurs because the tick-boxes in the userlist are not shown when the administrator has only EditState permission.

#### System information (as much as possible)

Any Joomla version from Joomla 2.5.28 until 3.4.1

Solution:
patch administrator/components/com_users/views/users/tmpl/default.php
in line 94:
old code in Joomla 3.4.1: <?php if ($canEdit) : ?>
New: <?php if ($canEdit || $canChange) : ?>

mbabker committed:
Mark tests skipped in PHP 7

mbabker committed:
Additional test fixes for PHP 7 dealing with deprecated MySQL driver

mbabker committed:
Implement a deprecation message handler for the unit test library

Based on Symfony's PhpUnit bridge, this handler will catch E_USER_DEPRECATED and E_DEPRECATED messages triggered while running the unit test suite and display them at the end of the test output.  This removes the need to silence deprecation messages in the unit test bootstrap and gives contributors an idea of deprecated code uses either in the core PHP platform or libraries.

Implementing this exposed a call to a non-static method in a static manner which only affects the unit test suite, this is also corrected.

Commit:
Merge pull request #7171 from mbabker/deprecationHandler

Implement a deprecation message handler for the unit test library

RickR2H committed:
[imp] Added switches to the admin module mod_status. Closes #7174.

bertmert committed:
[imp] Adds additional classes to pagination layout. Changes <i> to <span>. Closes #7175

infograf768 committed:
Adding missing formfield for PR #6718

wilsonge committed:
Fix the since tags to be 3.5

wilsonge committed:
Merge branch 'staging' into 3.5-dev

mbabker committed:
Deprecate JPlatform

wilsonge committed:
Merge pull request #7188 from mbabker/DeprecateJPlatform

Deprecate JPlatform

zero-24 committed:
Disable the new  option per default Fixes #7195

mbabker committed:
Change PHP 7 Exception handling to catch Throwable

Commit:
Merge pull request #7198 from mbabker/Throwable

Change PHP 7 Exception handling to catch Throwable

mbabker committed:
Define JVersion class vars as constants

Commit:
Merge pull request #7217 from mbabker/JVersionConst

Define JVersion class vars as constants

mbabker committed:
Merge remote-tracking branch 'joomla/staging' into 3.5-dev

smehrbrodt committed:
Improve wording: Purge -> Clear Fixes #7380

Commit:
Adding Unicode Supplementary Characters support in Joomla!

Fixes #7173

wilsonge committed:
JSession refactor Fixes #5088

fastslack committed:
Update JFormFieldSQL to use JDatabaseQuery (Fix #3476)

roland-d committed:
[imp] Add the ability to use a custom CSS file with Protostar (Fix #4211)

wilsonge committed:
Convert Gmail authentication plugin to use JHttp (Fix #6692)

dgrammatiko committed:
Remove progressbar.js dependency from com_joomlaupdate (Fix #6769)

AteDev committed:
Add anchor CSS option for menu item type of separator (Fix #6897)

smz committed:
Disable JS generation in JHtmlBoostrap::modal() (Fix #6919)

pe7er committed:
Featured Items view - added Alias + Hits (Fix #6949)

pe7er committed:
Implement Search Tools for Tags Manager (Fix #6950)

pe7er committed:
Implement Search Tools for News Feed Manager (Fix #6952)

Commit:
Added method getGroupTitle to JAccess (Fix #6994)

infograf768 committed:
Updatesites manager: Changing Publish to Enable (Fix #7154)

mbabker committed:
Merge remote-tracking branch 'joomla/staging' into 3.5-dev

mbabker committed:
Bump version for branch

mbabker committed:
Change packaging script to access JVersion data via constants

mbabker committed:
Use new Travis-CI infrastructure

Commit:
Merge branch 'staging' into 3.5-dev

Conflicts:
	administrator/language/en-GB/en-GB.mod_status.ini

mbabker committed:
Composer update for Joomla Framework packages and PHPUnit

mbabker committed:
Merge remote-tracking branch 'joomla/staging' into 3.5-dev

wilsonge committed:
Merge pull request #7524 from mbabker/3.5-Composer-Updates

Composer update for Joomla Framework packages and PHPUnit

Hackwar committed:
Fixing bug in standard routing behavior

Hackwar committed:
Merge branch '3.5-dev' of https://github.com/joomla/joomla-cms into componentrulesstandard

Conflicts:
	tests/unit/suites/libraries/cms/router/JRouterSiteTest.php

Hackwar commented:
I've combined the changes from this and all other routing related PRs into a new PR: #7615 Please review and comment in the new PR. I'm closing this one, so that we can focuse on the new PR.
